File: examples/2015-day-01/python/main.py
from pathlib import Path


def step(char: str) -> int:
    match char:
        case "(":
            return 1
        case ")":
            return -1
        case _:
            return 0


def solve_part_one(data: str) -> int:
    floor = 0

    for char in data:
        floor += step(char)

    return floor


def solve_part_two(data: str) -> int:
    floor = 0

    for index, char in enumerate(data, start=1):
        floor += step(char)

        if floor == -1:
            return index

    return 0


def main() -> None:
    data = Path("input.txt").read_text(encoding="utf-8")

    print(solve_part_one(data))
    print(solve_part_two(data))


if __name__ == "__main__":
    main()

File: examples/2015-day-01/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("failed to read input.txt");

    let part_one_solution = solve_part_one(&input);
    let part_two_solution = solve_part_two(&input);

    println!("{part_one_solution}");
    println!("{part_two_solution}");
}

fn step(floor: i64, instruction: char) -> i64 {
    match instruction {
        '(' => floor + 1,
        ')' => floor - 1,
        _ => floor,
    }
}

fn solve_part_one(input: &str) -> i64 {
    input.chars().fold(0, step)
}

fn solve_part_two(input: &str) -> i64 {
    let mut floor = 0;

    for (index, instruction) in input.chars().enumerate() {
        floor = step(floor, instruction);

        if floor == -1 {
            return index as i64 + 1;
        }
    }

    0
}

#[cfg(test)]
mod tests {
    use crate::{solve_part_one, solve_part_two};

    #[test]
    fn solves_part_one() {
        let cases = [
            ("(())", 0),
            ("()()", 0),
            ("(((", 3),
            ("(()(()(", 3),
            ("))(((((", 3),
            ("())", -1),
            ("))(", -1),
            (")))", -3),
            (")())())", -3),
        ];

        for (input, expected) in cases {
            let actual = solve_part_one(input);

            assert_eq!(actual, expected, "failed on input \"{input}\"")
        }
    }

    #[test]
    fn solves_part_two() {
        let cases = [(")", 1), ("()())", 5)];

        for (input, expected) in cases {
            let actual = solve_part_two(input);

            assert_eq!(actual, expected, "failed on input \"{input}\"")
        }
    }
}
